### PR TITLE
Check dir type for changes

### LINKED ISSUE #118: No change detected when altering plugin URI from upstream repo to local path
Hello,

Thanks again for your continued work on this project. I've been enjoying using it lately!

There seems to be an issue when changing between local projects and upstream URLs. For example:

1. Specify an upstream plugin.
	```lua
	return require('packer').startup{function(use)
		use {'rust-lang/rust.vim', opt=true}
	end}
	```
2. Install the upstream plugin.
	```sh
	nvim +PackerCompile +PackerSync
	```
3. Re-specify the plugin to use a local path.
	```lua
	return require('packer').startup{function(use)
		use {fn.glob('~/Programming/ACTIVE/rust.vim'), opt=true}
	end}
	```
4. Install the local plugin.
	```sh
	nvim +PackerCompile +PackerSync
	```

The second `PackerSync` will report "already clean" and won't replace the currently downloaded plugin with a symlink to the directory specified.

### PR DESCRIPTION
**Pull request by Iron-E**
This is a followup to #143. It uses a more consistent method for checking the type of a directory than attempting to embed it into the `packer_compile`.

The following changes have been made:

* `plugin_utils.find_missing_plugins` now uses two metrics to determine whether or not a plugin is "missing":
    1. The plugin is not installed at all (directory is missing).
    2. The plugin is installed, but the `guess_dir_type` does not match `plugin.type`.
* `PackerClean` has been re-implemented in terms of `find_missing_plugins`. This is because some missing plugins are also plugins which need to be cleaned before a reinstall (i.e. moving from local to upstream and vice versa).
    * See the commit message for 51bf0962d450eebe7d123a41279810ee1724406c. Essentially it just uses more logic from other parts of the plugin.
* More strings and numbers have been converted to variables in case the values change in the future.

Closes #118.

### CODE REVIEW

File: lua/packer/clean.lua
Comment (Iron-E) [inline lua/packer/clean.lua:57]: I have to add `if not plugin_source` here as well.

File: lua/packer/clean.lua
Comment (Iron-E) [inline lua/packer/clean.lua:54]: I'm going to add something here that removes detected plugins from the list, so that anything left over must not have been in the plugin spec and should be removed.

File: lua/packer.lua
Comment (wbthomason) [inline lua/packer.lua:181]: Nitpick: I'd request that `custom_plugin` be named `custom_plugin_type`.

File: lua/packer/clean.lua
Comment (wbthomason) [inline lua/packer/clean.lua:52]: Nitpick: Here and at line 59, is there a reason this isn't `table.insert(dirty_plugins, path)`?

File: lua/packer/display.lua
Comment (wbthomason) [inline lua/packer/display.lua:225]: Nitpick: Would request that `git_plugin` be called `git_plugin_type`.

File: lua/packer/plugin_utils.lua
Comment (wbthomason) [inline lua/packer/plugin_utils.lua:12]: Nitpick: would request that `local_plugin` be called `local_plugin_type`.

**Review by wbthomason — CHANGES_REQUESTED** [review]
Thanks! This looks like a significant improvement. I've left a few requests for minor changes (mostly just variable naming), but will be happy to merge.

We will eventually need to work this into `packer.install` and `packer.update` too, since not everyone runs `packer.sync` to ensure that the `clean` operation happens first (and - correct me if I'm wrong - the clean operation **must** run first for directory type changes to be picked up). Perhaps this is better in a different PR?

....to be honest, I'm now starting to wonder if it's better to just remove the `PackerInstall` and `PackerUpdate` commands entirely in lieu of `PackerSync` as the one-stop shop. What do you think? This change would not go in this PR, but thinking about the necessary changes to `install` and `update` has me feeling that there's some redundancy which could be removed there.

File: lua/packer/clean.lua
Comment (Iron-E) [reply lua/packer/clean.lua:52]: I usually try to use the insertion syntax where possible, because it's just one less resolution for Lua to go through (see [here](https://stackoverflow.com/questions/38345582/why-is-tt-1-e-faster-than-table-insertt-e)). I did notice that `table.insert` is preferred by a lot of the Neovim community, though, and I don't mind changing this— it was just my instinct.

**Comment by Iron-E** [comment]
I made the requested changes on the `xxx_plugin` variables. I see though that there is a `plugin_types` file— should these variables go there or stay in `plugin_utils`?

File: lua/packer/clean.lua
Comment (wbthomason) [reply lua/packer/clean.lua]: Oh, cool. I didn't know that - intuitively, I thought that the insertion syntax would trigger metamethods, whereas `table.insert` would not, so I actually thought `table.insert` would be faster. Good to learn, thanks! I'm fine with using the insertion syntax (or not if you don't feel like reverting the change to `table.insert`).

**Comment by cloggier** [comment]
Only downside in removing those commands is that when wanting to install new plugin or clean old one Sync would also run update.

**Comment by Iron-E** [comment]
Maybe `:PackerInstall` should always clean before installation. The user could choose to either clean if a change is made to the plugin type, but if the user doesn't want to (or if false positives occur) they can elect not to.

Then `:PackerSync` would just be `:PackerInstall` followed by `:PackerUpdate`, and the cleaning would happen during the installation routine.

**Comment by wbthomason** [comment]
> :PackerInstall also sees that a directory change is present but won't actually clean the existing path before attempting to install. This of course means that the :PackerInstall routine will yield failures when directory type changes are detected, since a cleaning is needed in that case.

Right - sorry, I wasn't very clear in my initial response; this is what I meant w.r.t. working the changes into `install`. This will probably also be an issue for `update` for the same reason.

**Review by wbthomason — APPROVED** [review]
This looks great, thank you! I think we should eventually refactor the `X_plugin_type` variables to `plugin_types`, but for this PR, I'm content for them to live in `plugin_utils`. I'm good to merge if you are. Thanks again!

File: lua/packer/clean.lua
Comment (Iron-E) [reply lua/packer/clean.lua]: I'll leave it to match the style in the rest of the project, though we can create another performance PR if we wanted to use that syntax for appending.

**Comment by Iron-E** [comment]
Looks good to me.

Glad I could help with this, thanks for reviewing!

**Pull request merged**
